src/parcsr_mv/par_csr_triplemat_device.c

```diff
@@ -859,6 +856,40 @@ hypre_ParCSRMatrixRAPKTDevice( hypre_ParCSRMatrix *R,
    hypre_ParCSRMatrixCompressOffdMapDevice(C);
    hypre_ParCSRMatrixCopyColMapOffdToHost(C);
 
+   /* Ensure that the diagonal entries exist in the matrix structure (even if numerically zero) */
+   if (hypre_CSRMatrixCheckForMissingDiagonal(C_diag))
+   {
+      hypre_CSRMatrix *zero_diag = hypre_CSRMatrixIdentityDevice(hypre_ParCSRMatrixNumRows(C), 0.0);
+      hypre_CSRMatrix *zero_offd = hypre_CSRMatrixCreate(hypre_CSRMatrixNumRows(zero_diag), 0, 0);
+      hypre_CSRMatrixInitialize_v2(zero_offd, 0, HYPRE_MEMORY_DEVICE);
+
+      hypre_ParCSRMatrix *zero = hypre_ParCSRMatrixCreate(hypre_ParCSRMatrixComm(A),
+                                                          hypre_ParCSRMatrixGlobalNumRows(C),
+                                                          hypre_ParCSRMatrixGlobalNumCols(C),
+                                                          hypre_ParCSRMatrixRowStarts(C),
+                                                          hypre_ParCSRMatrixColStarts(C),
+                                                          0,
+                                                          hypre_CSRMatrixNumNonzeros(zero_diag),
+                                                          hypre_CSRMatrixNumNonzeros(zero_offd));
+
+      hypre_CSRMatrixDestroy(hypre_ParCSRMatrixDiag(zero));
+      hypre_ParCSRMatrixDiag(zero) = zero_diag;
+
+      hypre_CSRMatrixDestroy(hypre_ParCSRMatrixOffd(zero));
+      hypre_ParCSRMatrixOffd(zero) = zero_offd;
+
+      hypre_ParCSRMatrix *C_new;
+      hypre_ParCSRMatrixAdd(1.0, C, 1.0, zero, &C_new);
```

Comment:
Thinking again, would it be better to just use `hypre_CSRMatrixAdd` on the diagonal portion instead? This way we don't need `zero_offd` and the ParCSRMatrix.

Reply:
Yes, I agree this is the correct way to do it, particularly since the check for missing diagonals is local to each rank, so not all ranks may enter the code branch guarded by the if statement here.